File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/math_trig.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import kotlin.math.*

suspend fun Piper.executeMathTrig(statement: JSONObject, onError: (e: Exception) -> Unit): Any? {
    try {
        Log.d("hello", "execute_math_trig")
        val pointer = statement.getJSONObject("pointer")
        val op = pointer["op"].toString()
        val value = executeStatement(pointer["value"].toString())

        val valueDouble = value.toString().toDouble()

        return when (op) {
            "SIN" -> sin(valueDouble / 180 * Math.PI)
            "COS" -> cos(valueDouble / 180 * Math.PI)
            "TAN" -> tan(valueDouble / 180 * Math.PI)
            "ASIN" -> asin(valueDouble) / Math.PI * 180
            "ACOS" -> acos(valueDouble) / Math.PI * 180
            "ATAN" -> atan(valueDouble) / Math.PI * 180
            else -> null
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}

File: piper/extensions/chrome/src/engine/statements/math_trig.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let op = stringify(pointer.op)
        let value = await this.executeStatement(stringify(pointer.value))

        switch(op) {
            case "SIN":
                return Math.sin(value / 180 * Math.PI)
            case "COS":
                return Math.cos(value / 180 * Math.PI)
            case "TAN":
                return Math.tan(value / 180 * Math.PI)
            case "ASIN":
                return Math.asin(value) / Math.PI * 180
            case "ACOS":
                return Math.acos(value) / Math.PI * 180
            case "ATAN":
                return Math.atan(value) / Math.PI * 180
        }
    } catch (e) {
        onError(e)
    }

    return null
}
